### DISCUSSION TITLE
Run in coroutine context

### DISCUSSION THREAD

**Discussion by ivsokol** [komapper/komapper#879] [Q&A] [answered]
Hi,
I am relatively new to Kotlin landscape and I am trying to pick ORM library for server side development. I already tried several libraries and was dissatisfied with their handling of joins, but in Komapper it just works :)

But what I wanted to ask is how to achieve running queries in couroutine IO context. Basically I am trying to achieve something like this in Komapper JDBC (in Ktor it is done with Exposed)

https://ktor.io/docs/interactive-website-add-persistence.html#queries 

`
suspend fun <T> dbQuery(block: suspend () -> T): T =
        newSuspendedTransaction(Dispatchers.IO) { block() }
`

Sorry for asking question about coroutines in Komapper discussion, but I was hoping I could get here some guidelines. I am also ok with switching to R2DBC if that is needed.

Kind regards,
Ivan

**Comment by nakamura-to** [comment, marked as answer]
Hi, Thanks for posting.

It is challenging to achieve JDBC access in a coroutine IO context because handling thread-local managed transactions in coroutines takes a lot of work.

There are some issues with the newSuspendedTransaction method.
https://github.com/JetBrains/Exposed/issues?q=is%3Aissue+is%3Aopen+newSuspendedTransaction

I recommend R2DBC if you want to use Komapper in combination with coroutines.

Alternatively, consider giving up coroutines and using JDBC, which is more stable than R2DBC.

**Comment by ivsokol** [comment]
Hi,

thank you for the quick response. I replaced jdbc with r2dbc and found appropriate methods, now everything works.

Kind regards,
Ivan

**Comment by ivsokol** [comment]
Hi,

I have one additional question. How to use FlowQuery with include() or includeAll() methods? Are they available at all for FlowQuery?

Regular query works and it maps result to EntityStoreQuery, but I cannot build FlowQuery. As soon as I add include, it doesn't work

**Reply by nakamura-to** [reply]
Hi,

include() and includeAll() methods are not available for FlowQuery.
We believe that we never want to use Flow to represent association navigation.
But can you tell us your use case?

**Reply by ivsokol** [reply]
Then can you advise what to do for leftJoin and flow result? Is this possible

My code looks like this (obb and obv are Meta from corresponding tables and I am reading rows from obb table where status is PENDING)

` 
val obb = Meta.onebbdao
val obv = Meta.onebbVoiceDAO
val query :EntityStoreQuery = QueryDsl.from(obb).leftJoin(obv) { obb.id eq obv.oneBBId }
            .where { obb.wwmsStatus eq WWMSStatusEnum.PENDING }
            .include(obb, obv)

val flowQuery: FlowQuery<???> = QueryDsl.from(obb)
            .leftJoin(obv) { obb.id eq obv.oneBBId }
            .where { obb.wwmsStatus eq WWMSStatusEnum.PENDING }
            .include(obb,obv)
`

flowQuery can be built until I add include() command.

Basically, I want to get result set from joined tables as flow and process it like that

**Reply by nakamura-to** [reply]
> Then can you advise wat to do for leftJoin and flow result? Is this possible

You can write it as follows:

```kotlin
val storeFlow: Flow<EntityStore> = flow {
    val store = db.runQuery {
        QueryDsl.from(obb)
        .leftJoin(obv) { obb.id eq obv.oneBBId }
        .where { obb.wwmsStatus eq WWMSStatusEnum.PENDING }
        .include(obb,obv)
    }
    emit(store)
}
```

You can also use [Transactional flow](https://www.komapper.org/docs/reference/transaction/#transactional-flow) to represent the entire transaction as a flow:

```kotlin
val storeFlow: Flow<EntityStore> = db.flowTransaction {
    val store = db.runQuery {
        QueryDsl.from(obb)
        .leftJoin(obv) { obb.id eq obv.oneBBId }
        .where { obb.wwmsStatus eq WWMSStatusEnum.PENDING }
        .include(obb,obv)
    }
    emit(store)
}
```

**Reply by ivsokol** [reply]
thanks, I would probably then do that. I thought this could be resolved somehow in komapper, so that komapper emits flow events on each row read

**Reply by nakamura-to** [reply]
A Flow obtained from `R2dbcDatabase.flowQuery` represents the rows connected to the database.
Note, however, that an EntityStore is a set of rows disconnected from the database, and treating the EntityStore as a Flow is not efficient. That is why Komapper does not provide a way to get a Flow<EntityStore>.
